src/net/server/handlers/query.rs

```diff
@@ -64,13 +65,36 @@ impl<T> Clone for ReservedPermit<T> {
 
 /// Middleware that first reserves a permit on the channel to send messages to the messaging layer.
 /// Once reserved, adds the permit to the extension for retrieval from the handler.
-pub async fn obtain_permit_mw<T: Send + 'static, B>(
-    sender: mpsc::Sender<T>,
-    mut req: Request<B>,
+/// # Panics
+/// if messages arrive out of order
+pub async fn obtain_permit_mw<B: Send>(
+    req: Request<B>,
     next: Next<B>,
 ) -> Result<Response, MpcHelperServerError> {
+    let mut req_parts = RequestParts::new(req);
+    let Path(query_id, step) = Path::from_request(&mut req_parts).await?;
+    let Query(RoleQueryParam { role }) =
+        Query::<RoleQueryParam>::from_request(&mut req_parts).await?;
+    let record_headers = RecordHeaders::from_request(&mut req_parts).await?;
+    let Extension(last_seen_messages) =
+        Extension::<LastSeenMessages>::from_request(&mut req_parts).await?;
+    // PANIC if messages arrive out of order
+    // TODO (ts): remove this when streaming solution is complete
+    last_seen_messages
+        .update_in_place(ChannelId::new(role, step), record_headers.offset)
+        .unwrap();
```

Comment:
why `unwrap` instead of `?`

Reply:
I thought I should panic when I discover out-of-order data. would we rather just return a failure?

Reply:
`update_in_place` returns the same error type as this function, so I think it is a matter of changing `unwrap` to `?`. if you want to panic, you would probably want to change `update_in_place` definition

Reply:
done